src/vmm/src/lib.rs

```diff
@@ -1270,7 +1270,26 @@ impl Vmm {
             match self.epoll_context.dispatch_table[event.data as usize] {
                 Some(EpollDispatch::Exit) => {
                     self.exit_evt.read().map_err(Error::EventFd)?;
-                    self.stop(i32::from(FC_EXIT_CODE_OK));
+
+                    // Query each vcpu for the exit_code.
+                    // If the exit_code can't be found on any vcpu, it means that the exit signal
+                    // has been issued by the i8042 controller in which case we exit with
+                    // FC_EXIT_CODE_OK.
+                    let exit_code = self
+                        .vcpus_handles
+                        .iter()
+                        .find_map(|handle| {
+                            match handle
+                                .response_receiver()
+                                .recv_timeout(Duration::from_millis(1))
```

Comment:
I think we could spare a few more milliseconds here :smile:

Reply:
+1 We actually hit this issue under heavy load due to testing (launching multiple VMs). 10d1f4aa59da50f8d39db9692844740dbacde5e6

Reply:
This is an error case and only impacts the return code. We can't be extra generous either and optimize for the error case because then we'd be hurting the common good case by delaying the actual process end by `num_vcpus * timeout`.

Reply:
We shouldn't need more than 1 since the vCPU first writes the response and then it writes to `exit_evt`. But I can increase this number. How many would you suggest ?

Reply:
If `KVM_EXIT_SHUTDOWN` is not returned (as we've seen on certain cpu/kernel-version combinations) we're only relying on the `i8042` exit event which triggers the `Vmm` exit code without having any `Vcpu` write any exit reason. It's a valid case explained by your comment.

Reply:
@serban300 is right, no need for bigger timeout.

1. in case of vcpu exit/error: `Vcpu` populates the channel **then** hits the event-fd, so by the time vmm picks up the event the channel data is guaranteed to be there.
2. in case of `i8042`: `Vcpu` doesn't provide data on the channel and Vmm will simply find it empty.

In all cases we don't need to wait.

nit: replace `.recv_timeout()` with `.try_wait()`.

Reply:
Fixed.